Optimizer Sanity Checker

yfy- opened the pull request:
## Rationale for this change

We extend the existing `PipelineChecker` with additional checks of order and distribution requirements. In a physical plan each plan should be pipeline friendly (previously checked by `PipelineChecker`) and each child plan's output order and partitioning should satisfy the parent's respective requirements. We combine these checks into the `SanityCheckPlan` proposed with this PR.

With the `SanityCheckPlan`, as the optimizer steps change and grow, we ensure that the order and distribution requirements of the final phsyical plan are always satisfied so that it can yield correct results.

## What changes are included in this PR?

1. The `SanityCheckPlan` step as the last step of the physical plan optimizations.
2. `SanityCheckPlan` contains the former `PipelineChecker`, that's why `PipelineChecker` is deleted.

## Are these changes tested?

We use the existing test cases of the former `PipelineChecker`. In addition following cases are added:

#### Using `BoundedWindowAggExec`
2 cases: First with the child order requirement is satisfied and another one that is not satisifed.

#### Using `GlobalLimitExec`
2 cases: First with the child distribution requirement is satisfied and another one that is not satisifed.

#### Using `LocalLimitExec`
We check when there are no requirements at all our check passes.

#### Using `SortMergeJoinExec`
3 cases: First case with both children satisify both requirements. Second case, where the second child does not satisfy the order requirement. Finally, a case where the second child does not satisfy distribution requirements.

yfy- committed:
Initial optimizer sanity checker.

Only includes sort reqs, docs will be added.

mustafasrepo commented on datafusion/core/src/physical_optimizer/sanity_checker.rs:
```diff
@@ -0,0 +1,152 @@
+        let child_sort_req = child_sort_req.as_ref().unwrap();
+        if !child
+            .equivalence_properties()
+            .ordering_satisfy_requirement(&child_sort_req)
```
This API is the correct API (in response to your email).

mustafasrepo commented on datafusion/core/src/physical_optimizer/sanity_checker.rs:
```diff
@@ -0,0 +1,152 @@
+        let bw = bounded_window_exec("c9", sort_exprs, sort);
+        let sanity_checker = SanityCheckPlan::new();
+        let opts = ConfigOptions::default();
+        assert_eq!(sanity_checker.optimize(bw, &opts).is_ok(), true);
```
Additionally you can assert plan of the `bw` to make test more explicit. By this way plan will be more visible: Somethinkg like below:
```rust
let expected_plan = vec![
 // Fill Plan
];
let actual_plan = get_plan_string(&bw);
assert_eq(expected_plan, actual_plan)
```

yfy- replied:
I have added plan checks in the below commit:

https://github.com/synnada-ai/datafusion-upstream/pull/24/commits/aea95c66f7fe59826ea435597b2009a24bebf897

mustafasrepo commented on datafusion/core/src/physical_optimizer/sanity_checker.rs:
```diff
@@ -0,0 +1,152 @@
+                nulls_first: false,
+            },
+        )];
+        let bw = bounded_window_exec("c9", sort_exprs, source);
```
Same thing applies here

yfy- replied:
I have added plan checks in the below commit:

https://github.com/synnada-ai/datafusion-upstream/pull/24/commits/aea95c66f7fe59826ea435597b2009a24bebf897

mustafasrepo replied:
👍

yfy- committed:
Add distro and pipeline friendly checks

yfy- committed:
Also check the plans we create are correct.

berkaysynnada commented on datafusion/core/src/physical_optimizer/optimizer.rs:
```diff
@@ -124,6 +125,10 @@ impl PhysicalOptimizer {
+            // SanityChecker rule checks whether the order and the
+            // partition requirements of each execution plan is
+            // satisfied by its children or not.
+            Arc::new(SanityCheckPlan::new()),
```
We can discuss if this rule can be merged with `PipelineChecker`.

yfy- replied:
From my understanding, in addition to doing `is_pipeline_friendly`, `PipelineChecker` also checks a condition around `SymmetricHashJoin` operation. Since `PipelineChecker` is more enchanced (to check pipeline friendliness) should we remove the pipeline friendliness check from `SanityChecker`? Or to simplify, I can merge `PipelineChecker` into `SanityChecker` (and remove `PipelineChecker`).

mustafasrepo replied:
I think, we should either 
- remove pipeline friendliness check from `SanityChecker` (as this is done in another rule already)
- or merge `PipelineChecker` into `SanityChecker` .

However, merging these two rules seems more appropriate. Hence, we should follow this approach.

yfy- replied:
They are merged in the below commit:

https://github.com/synnada-ai/datafusion-upstream/pull/24/commits/d89b1c91c9d8e8571dd2757393ed6fe7e0619cdf

berkaysynnada commented on datafusion/core/src/physical_optimizer/sanity_checker.rs:
```diff
@@ -0,0 +1,175 @@
+    }
+}
+
+/// Ensures that the plan is pipeline friendly and the order and
```
It does not actually measure the pipeline friendliness.

berkaysynnada commented on datafusion/core/src/physical_optimizer/sanity_checker.rs:
```diff
@@ -0,0 +1,175 @@
+    plan: Arc<dyn ExecutionPlan>,
+) -> Result<Transformed<Arc<dyn ExecutionPlan>>> {
+    if !plan.execution_mode().pipeline_friendly() {
+        return plan_err!("Plan {:?} is not pipeline friendly.", plan);
```
Pipeline friendly means the plan can be executed in bounded memory even if the source is unbounded, or the plan has a bounded source. If we want to apply this rule before `PipelineChecker`, we need to remove this check, or alternatively we can move this rule after `PipelineChecker`.

yfy- committed:
Add distribution test cases using global limit exec.

yfy- commented:
@mustafasrepo, about the testing strategy: So far I have picked one scenario for order requirements (using `BoundedWindowAggExec`), one scenario for partition requirements (`GlobalLimitExec`) and a  scenario when there are no requirements at all (`LocalLimitExec`).

Should we extend this to cover all `ExecutionPlan`'s that have some requirement, sort and/or distribution? I can add test cases for the following as well:

- `AggregateExec`
- `RecursiveQueryExec` 
- `SymmetricHashJoinExec`
- `NestedLoopJoinExec`
- `SortMergeJoinExec`
- `CrossJoinExec`
- `HashJoinExec`
- `DataSinkExec`
- `WindowAggExec`
- `SortExec`
- `PartialSortExec`
- `SortPreservingMergeExec`

mustafasrepo commented:
> @mustafasrepo, about the testing strategy: So far I have picked one scenario for order requirements (using `BoundedWindowAggExec`), one scenario for partition requirements (`GlobalLimitExec`) and a scenario when there are no requirements at all (`LocalLimitExec`).
> 
> Should we extend this to cover all `ExecutionPlan`'s that have some requirement, sort and/or distribution? I can add test cases for the following as well:
> 
> * `AggregateExec`
> * `RecursiveQueryExec`
> * `SymmetricHashJoinExec`
> * `NestedLoopJoinExec`
> * `SortMergeJoinExec`
> * `CrossJoinExec`
> * `HashJoinExec`
> * `DataSinkExec`
> * `WindowAggExec`
> * `SortExec`
> * `PartialSortExec`
> * `SortPreservingMergeExec`

I think, existing test coverage is good. We do not need to use all operators in the tests. However, all the operators in the tests are single child operators. We can also add a test case for an operator with multiple children (such as `SortMergeJoinExec`, it has both input distribution requirement and order requirement. It will allow us to test rule with full coverage.) Once we add these test cases. I think test coverage will be in very good shape, and will be sufficient.

mustafasrepo commented:
Also, I ran the CI tests. It seems that we have some clippy errors and test errors (I think they are related to pipeline error raised in the Sanity Checker Rule. Since error message is changed expected error message isn't same anymore. Once you merge PipelineChecker rule and Sanity Checker rule these errors will be fixed as far as I can tell).
Make sure to run command
`ulimit -S -n 2048 cargo test` (for mac)
`cargo test` (for other systems)
in the `.git` main directory to see whether all tests pass. This test may take 3-5 minutes.

Also, CI tests has additional check whether code follows guidelines. For this check, we have the script `bash dev/rust_lint.sh`
This script checks the style of the codebase. It also guides you, how to conform to the guide. Once `bash dev/rust_lint.sh`, and `ulimit -S -n 2048 cargo test` checks complete successfully. CI tests will pass 99% of the time.

If you encounter any issues during these stages, please notify me. I can help you out.

yfy- committed:
Add test for multiple children using SortMergeJoinExec.

yfy- committed:
Move PipelineChecker to SanityCheckPlan

yfy- committed:
Fix some tests and add docs

yfy- committed:
Add some test docs and fix clippy diagnostics.

yfy- commented:
> > @mustafasrepo, about the testing strategy: So far I have picked one scenario for order requirements (using `BoundedWindowAggExec`), one scenario for partition requirements (`GlobalLimitExec`) and a scenario when there are no requirements at all (`LocalLimitExec`).
> > Should we extend this to cover all `ExecutionPlan`'s that have some requirement, sort and/or distribution? I can add test cases for the following as well:
> > 
> > * `AggregateExec`
> > * `RecursiveQueryExec`
> > * `SymmetricHashJoinExec`
> > * `NestedLoopJoinExec`
> > * `SortMergeJoinExec`
> > * `CrossJoinExec`
> > * `HashJoinExec`
> > * `DataSinkExec`
> > * `WindowAggExec`
> > * `SortExec`
> > * `PartialSortExec`
> > * `SortPreservingMergeExec`
> 
> I think, existing test coverage is good. We do not need to use all operators in the tests. However, all the operators in the tests are single child operators. We can also add a test case for an operator with multiple children (such as `SortMergeJoinExec`, it has both input distribution requirement and order requirement. It will allow us to test rule with full coverage.) Once we add these test cases. I think test coverage will be in very good shape, and will be sufficient.

I have added test using `SortMergeJoinExec` in the below commit:

https://github.com/synnada-ai/datafusion-upstream/pull/24/commits/49aba4b0a541e6bb103f3aa4392589fc30298e5f

yfy- commented:
> Also, I ran the CI tests. It seems that we have some clippy errors and test errors (I think they are related to pipeline error raised in the Sanity Checker Rule. Since error message is changed expected error message isn't same anymore. Once you merge PipelineChecker rule and Sanity Checker rule these errors will be fixed as far as I can tell). Make sure to run command `ulimit -S -n 2048 cargo test` (for mac) `cargo test` (for other systems) in the `.git` main directory to see whether all tests pass. This test may take 3-5 minutes.
> 
> Also, CI tests has additional check whether code follows guidelines. For this check, we have the script `bash dev/rust_lint.sh` This script checks the style of the codebase. It also guides you, how to conform to the guide. Once `bash dev/rust_lint.sh`, and `ulimit -S -n 2048 cargo test` checks complete successfully. CI tests will pass 99% of the time.
> 
> If you encounter any issues during these stages, please notify me. I can help you out.

Thanks @mustafasrepo, I have fixed clippy errors and some other test cases. However, there is one test case that does not finish: `fifo::unix_test::unbounded_file_with_symmetric_join`. The query defined there gets stuck at physical plan optimization stage and it hits the check where we look at the distribution requirements in `SanityCheckPlan`. When I remove the distribution requirements check the tests execute successfully. When I examined that physical plan I observed `SymmetricHashJoinExec` requires its 2 children to be hash partitioned on join columns, but its 2 children have `UndefinedDistribution`. Somehow, `SanityCheckPlan` fails and it causes the query optimization to execute indefinitely (at least on my machine). I am further examining why this would happen.

mustafasrepo committed:
Fix some failing tests

mustafasrepo commented:
> Thanks @mustafasrepo, I have fixed clippy errors and some other test cases. However, there is one test case that does not finish: `fifo::unix_test::unbounded_file_with_symmetric_join`. The query defined there gets stuck at physical plan optimization stage and it hits the check where we look at the distribution requirements in `SanityCheckPlan`. When I remove the distribution requirements check the tests execute successfully. When I examined that physical plan I observed `SymmetricHashJoinExec` requires its 2 children to be hash partitioned on join columns, but its 2 children have `UndefinedDistribution`. Somehow, `SanityCheckPlan` fails and it causes the query optimization to execute indefinitely (at least on my machine). I am further examining why this would happen.

These problems do not seem to be related with your work. Hence, I also debugged them. It seems that we have two problems
- `fifo::unix_test::unbounded_file_with_symmetric_join` test stucks when error is returned. (it has some tasks running. those tasks couldn't be joined when error is returned. Then test doesn't finish.)
- Partitioning `UnknownDistribution(1)` doesn't satisfy requirement `HashRequirement(vec![expr1, ..])`. This requirement is actually satisfied (The root cause of the error). 

I sent a commit to fix these problems. Also there are some other tests failing with this check. After examining them, I have found that rule actually help us find invalid plans. In the commit I also fixed them. However, there are still some clippy errors (If you do not have those errors. They come with latest rust updates. You can do `rustup update` or corresponding command in your local to bring latest stable changes.) Also as far as I remember those clippy errors are already resolved in the upstream. Hence I recommend you to merge your branch with `apache_main` before resolving any clippy issues. They might already be solved.

mustafasrepo commented on datafusion/physical-plan/src/aggregates/mod.rs:
```diff
@@ -675,7 +675,7 @@ impl ExecutionPlan for AggregateExec {
             }
             AggregateMode::FinalPartitioned | AggregateMode::SinglePartitioned => {
-                vec![Distribution::HashPartitioned(self.output_group_expr())]
+                vec![Distribution::HashPartitioned(self.group_by.input_exprs())]
```
We should have required input expressions. Rule helped us to catch this bug.

mustafasrepo commented on datafusion/core/tests/tpcds_planning.rs:
```diff
@@ -1044,7 +1044,10 @@ async fn regression_test(query_no: u8, create_physical: bool) -> Result<()> {
-            Arc::new(MemTable::try_new(Arc::new(table.schema.clone()), vec![])?),
+            Arc::new(MemTable::try_new(
+                Arc::new(table.schema.clone()),
+                vec![vec![]],
```
At the source, we were generating empty partitions (0 partition). This is invalid. Rule helped us to catch this bug.

yfy- committed:
Merge branch 'apache_main' into optimizer-sanity-checker

yfy- committed:
Replace PipelineChecker with SanityChecker in .slt files.

yfy- commented:
> > Thanks @mustafasrepo, I have fixed clippy errors and some other test cases. However, there is one test case that does not finish: `fifo::unix_test::unbounded_file_with_symmetric_join`. The query defined there gets stuck at physical plan optimization stage and it hits the check where we look at the distribution requirements in `SanityCheckPlan`. When I remove the distribution requirements check the tests execute successfully. When I examined that physical plan I observed `SymmetricHashJoinExec` requires its 2 children to be hash partitioned on join columns, but its 2 children have `UndefinedDistribution`. Somehow, `SanityCheckPlan` fails and it causes the query optimization to execute indefinitely (at least on my machine). I am further examining why this would happen.
> 
> These problems do not seem to be related with your work. Hence, I also debugged them. It seems that we have two problems
> 
> * `fifo::unix_test::unbounded_file_with_symmetric_join` test stucks when error is returned. (it has some tasks running. those tasks couldn't be joined when error is returned. Then test doesn't finish.)
> * Partitioning `UnknownDistribution(1)` doesn't satisfy requirement `HashRequirement(vec![expr1, ..])`. This requirement is actually satisfied (The root cause of the error).
> 
> I sent a commit to fix these problems. Also there are some other tests failing with this check. After examining them, I have found that rule actually help us find invalid plans. In the commit I also fixed them. However, there are still some clippy errors (If you do not have those errors. They come with latest rust updates. You can do `rustup update` or corresponding command in your local to bring latest stable changes.) Also as far as I remember those clippy errors are already resolved in the upstream. Hence I recommend you to merge your branch with `apache_main` before resolving any clippy issues. They might already be solved.

I have merged `apache_main` to the branch and also ran `rustup update`, now clippy errors seem to be gone. However, there are other test cases that are failing. These come from `bin/sqllogictests.rs`. From what I could get there is a problem with queries like:

```sql
SELECT 1 num UNION ALL SELECT 2 num ORDER BY num;
```

This yields the physical plan:
```
SortPreservingMergeExec: [num@0 ASC NULLS LAST]
  UnionExec
    ProjectionExec: expr=[1 as num]
      PlaceholderRowExec
    ProjectionExec: expr=[2 as num]
      PlaceholderRowExec
```

Order requirement for the `SortPreservingMergeExec` is not satisifed.

mustafasrepo commented:
> I have merged apache_main to the branch and also ran rustup update, now clippy errors seem to be gone. However, there are other test cases that are failing.

@yfy- It seems that this failure is another case this check helped us catch a bug. I will fix this problem. I will let you know once this problem is solved.

mustafasrepo committed:
Initial commit

mustafasrepo committed:
Slt tests pass

mustafasrepo committed:
Resolve linter errors

mustafasrepo committed:
Minor changes

mustafasrepo committed:
Minor changes

mustafasrepo committed:
Minor changes

mustafasrepo committed:
Minor changes

mustafasrepo commented:
@yfy- I resolved the problems mentioned above in the [PR](https://github.com/yfy-/datafusion-upstream/pull/1). We may either merge that PR into your PR, or continue as separate PRs. (We will decide after internal discussion.) With the changes in that PR, all test pass in my local. Hence, for now we can call this PR complete in terms of functionality. However, I will do one last review of this PR tomorrow to make sure we are not missing anything. Kind regards

mustafasrepo committed:
Sort PreservingMerge clear per partition

mustafasrepo committed:
Minor changes

mustafasrepo committed:
Merge branch 'optimizer-sanity-checker' into sanity_checker_subs

berkaysynnada committed:
Update output_requirements.rs

mustafasrepo committed:
Address reviews

mustafasrepo committed:
Merge branch 'apache_main' into optimizer-sanity-checker

mustafasrepo committed:
Merge branch 'optimizer-sanity-checker' into sanity_checker_subs

mustafasrepo commented:
@yfy- this PR is ready to go upstream. 🎉 🎉 🚀

ozankabak commented on datafusion/core/src/physical_optimizer/optimizer.rs:
```diff
@@ -124,11 +124,15 @@ impl PhysicalOptimizer {
-            // diagnostic error message when this happens. It makes no changes to the
-            // given query plan; i.e. it only acts as a final gatekeeping rule.
-            Arc::new(PipelineChecker::new()),
+            // The SanityCheckPlan will reject non-runnable query
+            // plans that use pipeline-breaking operators on infinite
+            // input(s) as well as checks whether the order and
+            // distribution requirements of each children of a plan
+            // are satisfied. The rule generates a diagnostic error
+            // message when this happens. It makes no changes to the
+            // given query plan; i.e. it only acts as a final
+            // gatekeeping rule.
```
```suggestion
            // The SanityCheckPlan rule checks whether the order and
            // distribution requirements of each node in the plan
            // is satisfied. It will also reject non-runnable query
            // plans that use pipeline-breaking operators on infinite
            // input(s). The rule generates a diagnostic error
            // message for invalid plans. It makes no changes to the
            // given query plan; i.e. it only acts as a final
            // gatekeeping rule.
```

ozankabak commented on datafusion/physical-expr/src/equivalence/properties.rs:
```diff
@@ -917,7 +956,12 @@ impl EquivalenceProperties {
+            .constants
+            .iter()
+            .map(|const_expr| const_expr.expr().clone())
+            .collect::<Vec<_>>();
```
I don't think you need to collect here, `normalize_exprs` takes an iterator

ozankabak commented on datafusion/physical-plan/src/union.rs:
```diff
@@ -168,6 +138,67 @@ impl UnionExec {
     }
 }
 
+fn calculate_union_eq_properties(
```
Needs a docstring -- take a look at the typical docstring format we use

ozankabak commented on datafusion/core/src/physical_optimizer/sanity_checker.rs:
```diff
@@ -0,0 +1,670 @@
+
+use itertools::izip;
+
+/// The SanityCheckPlan rule rejects the following query plans:
+/// i) Plans that use pipeline-breaking operators on infinite input(s),
+///    these queries cannot be executed.
+/// ii) Plans in which their order and distribution requirements are not
+///     satisfied by its children, these queries will most likely yield
+///     incorrect results.
```
```suggestion
/// The SanityCheckPlan rule rejects the following query plans:
/// 1. Invalid plans containing nodes whose order and/or distribution requirements
///    are not satisfied by their children.
/// 2. Plans that use pipeline-breaking operators on infinite input(s),
///    it is impossible to execute such queries.
```

ozankabak commented on datafusion/core/src/physical_optimizer/sanity_checker.rs:
```diff
@@ -0,0 +1,670 @@
+        plan.required_input_distribution().iter()
+    ) {
+        let child_eq_props = child.equivalence_properties();
+        match child_sort_req {
+            None => (),
+            Some(child_sort_req) => {
+                if !child_eq_props.ordering_satisfy_requirement(child_sort_req) {
+                    let child_plan_str = get_plan_string(child);
+                    return plan_err!(
+                        "Child: {:?} does not satisfy parent order requirements: {:?}",
+                        child_plan_str,
+                        child_sort_req
+                    );
+                }
+            }
+        };
```
Why not use an `if let Some(child_sort_req) = ...`?

ozankabak reviewed (approved):
Left some minor comments to address, but LGTM in general. Great work @yfy-, please coordinate with @mustafasrepo to send/merge upstream after addressing the comments.

mustafasrepo committed:
Update datafusion/core/src/physical_optimizer/optimizer.rs

Co-authored-by: Mehmet Ozan Kabak <ozankabak@gmail.com>

mustafasrepo committed:
Update datafusion/core/src/physical_optimizer/sanity_checker.rs

Co-authored-by: Mehmet Ozan Kabak <ozankabak@gmail.com>

mustafasrepo committed:
Address reviews

mustafasrepo committed:
Merge branch 'apache_main' into optimizer-sanity-checker

mustafasrepo committed:
Minor changes

mustafasrepo commented:
@yfy- final reviews are addressed now. You can open this PR to upstream repo 🚀 🚀 .
